### ISSUE TITLE
osxphotos version 0.51.7 "no such table: config" export fails - please improve error output

### ISSUE THREAD

**Issue by rsr-at-mindtwin** [RhetTbull/osxphotos#794]
Please feel free to close this bug as is, but I wanted to create it to capture an issue so that other people can find it by searching, also I would suggest capturing a bit more error output when this issue occurs, especially in the crash log file - some relevant internal logic or sqlite error is getting swallowed in the code.

There seems to be some compatibility issue between the current version of osxphotos and previous versions with regard to the .osxphotos.db file created in the root directory of the export target directory.  The symptom is pretty straightforward, osxphotos will start, pause for awhile, and then exit with the following error;

`osxphotos version 0.51.7
exiftool path: /opt/homebrew/bin/exiftool
Using last opened Photos library: /Users/unm/Pictures/Photos Library.photoslibrary
Something went wrong and osxphotos encountered an error:
no such table: config
Crash log written to '/Users/unm/Downloads/o/dist/osxphotos_crash.log'
Please file a bug report at https://github.com/RhetTbull/osxphotos/issues with the crash log attached.`

Inspecting the crash log is useless, it simply dumps the settings and then:

`osxphotos version: 0.51.7
Error: no such table: config
...`

Note that commands like `./osxphotos info` will work just fine:

`% osxphotos info
Using last opened Photos library: /Users/rsr/Pictures/Photos Library.photoslibrary
database_path: /Users/rsr/Pictures/Photos Library.photoslibrary/database/photos.db
database_version: '6000'
photo_count: 28155...`

The issue here appears to be somethign about the .osxphotos_export.db file found in the export destination directory.

Interactively running "pragma integrity_check;" in sqlite opening the file does not report an error:

`% sqlite3 /Volumes/media2/Photos/unmPhonePhotos/Exported/COPIED_DEBUG.osxphotos_export.db
SQLite version 3.37.0 2021-12-09 01:34:53
Enter ".help" for usage hints.
sqlite> pragma integrity_check;
ok`

... so perhaps it is some schema change between the .4.x version I was running previously and the current version, or some lingering difference since my first export was _many_ versions ago.

The dumb way to clear this issue is to move the .db file aside, or remove it entirely since once you re-export it has no value. Perhaps there's a better way to deal with it.

**details:**

[osxphotos_crash.log](https://github.com/RhetTbull/osxphotos/files/9609238/osxphotos_crash.log)

pipx upgrade osxphotos (has issue)
uninstalled and used release osxphotos_MacOS_exe_darwin_x64_v0.51.7 (has issue)

version information:

`% ./osxphotos -v      
osxphotos, version 0.51.7`

OS is Monterey 12.6 (21G115) on an M1 mini.

**Comment by RhetTbull** [comment]
Migration code:

https://github.com/RhetTbull/osxphotos/blob/eedc2f0b059617dcd6809dc6603b4d96dc397071/osxphotos/export_db.py#L611-L763

**Comment by rsr-at-mindtwin** [comment]
Argh.  Unfortunately, I got really sloppy in my (further) experimentation and ended up nuking the export database. I actually just ended up re-exporting the whole library.   Sorry about that.

That said, I wonder if the issue is that the code around "# create config table" isn't being reached before the table is being accessed in all paths. When I dumped the schema in the exported db, the config table was absolutely not there.  

Well, hopefully it will help someone else to have this.

**Comment by RhetTbull** [comment]
Do you recall what version of osxphotos you were running before upgrading to the current version?  I might be able to recreate the issue.  Would definitely like to make sure osxphotos can be seamlessly upgraded.

**Comment by rsr-at-mindtwin** [comment]
Ah, idea. I have another machine that I also export from, a different account.  That machine had the same problem with 0.50.11, which I'd abandoned with the intent to come back later and fix it, so it cropped up before that version.  

[output.txt](https://github.com/RhetTbull/osxphotos/files/9610307/output.txt)

Unfortunately, I don't think this output is very revealing.  I will leave this other export db alone for the time being, but I have no idea what it was created with. Not older than the mavericks fixes, but not too much more recent than that.

**Comment by RhetTbull** [comment]
Thanks for the additional output. This is very odd as it appears your database is already at the current version but lacks one of the tables in the current version (`config`).  I tried to replicate this by installing `osxphotos==0.40.0` but when I run with the new version, the database does correctly get migrated:

`Upgraded export database /Users/rhet/Desktop/export/.osxphotos_export.db from version 3.2 to 7.1`

I'll keep poking around to see if I can figure out how this could happen.  If it's only the config table that is missing, you could try this, then try to re-export:

`sqlite3 /Volumes/media2/Photos/unm2PhonePhotos/Exported/.osxphotos_export.db "CREATE TABLE IF NOT EXISTS config(id INTEGER PRIMARY KEY, datetime TEXT, config TEXT);"`

**Comment by rsr-at-mindtwin** [comment]
Yeah, sorry I couldn't provide more info. I am usually pretty good at not upgrading unless there's a need but kind of lost the thread the other day. 

```
unm2@xM1 ~ % sqlite3 /Volumes/media2/Photos/unm2PhonePhotos/Exported/.osxphotos_export.db "CREATE TABLE IF NOT EXISTS config(id INTEGER PRIMARY KEY, datetime TEXT, config TEXT);"

~ %  /Users/unm2/.local/bin/osxphotos export --from-date "2021-12-01" --update --finder-tag-keywords --person-keyword --xattr-template findercomment "{title}{title?{descr?{newline},},}{descr}" --exiftool-merge-keywords --exiftool-merge-persons --exiftool --keyword-template "{label,nolabel}" --directory '{created.year}-{created.mm}/{place.name.city,unknown}' --filename '{created.year}-{created.mm}-{created.dd} {original_name}' --strip  /Volumes/media2/Photos/unm2PhonePhotos/Exported/
Using last opened Photos library: /Users/unm2/Pictures/Photos Library.photoslibrary
Exporting 7121 photos to /Volumes/media2/Photos/unm2PhonePhotos/Exported...
Error exporting photo (BC159616-F1D5-4DC5-BE85-934BE29F04F4: IMG_5426.HEIC) as 2022-03-20 IMG_5426.HEIC: no such table: photoinfo
Error exporting photo (FF45E070-D638-4982-873E-4B82E6403E28: IMG_6691.HEIC) as 2022-04-25 IMG_6691.HEIC: no such table: photoinfo
Error exporting photo (AE4FA718-3B33-4410-9B0C-11779458F04F: IMG_5506.HEIC) as 2022-03-21 IMG_5506.HEIC: no such table: photoinfo
Error exporting photo (1EF0A43B-F82B-4578-BB70-004AD912D373: IMG_1723.HEIC) as 2022-08-25 IMG_1723.HEIC: no such table: photoinfo
Error exporting photo (006CC387-CF7D-4745-9A08-50C487B5B214: IMG_9100.HEIC) as 2022-06-23 IMG_9100.HEIC: no such table: photoinfo
Error exporting photo (3D34A155-2FF1-48D7-A455-6F5275265310: IMG_1748.HEIC) as 2022-08-25 IMG_1748.HEIC: no such table: photoinfo
Error exporting photo (107B1B3D-5041-47F0-9B7F-5893D5B532F8: IMG_1089.HEIC) as 2022-08-05 IMG_1089.HEIC: no such table: photoinfo
Error exporting photo (C56280F8-91D6-4832-A38F-3488D5187BBC: IMG_8605.HEIC) as 2022-06-23 IMG_8605.HEIC: no such table: photoinfo
Error exporting photo (0BF785BC-6FB0-48E6-86AE-636101DE5B2E: IMG_7806.HEIC) as 2022-06-14 IMG_7806.HEIC: no such table: photoinfo
Error exporting photo (A7AD90ED-0D84-4E06-A5D6-8ACF377FC02A: IMG_7214.HEIC) as 2022-05-30 IMG_7214.HEIC: no such table: photoinfo
Error exporting photo (C2DA3377-F5D3-46D8-AE40-8D8D1C312759: IMG_7039.HEIC) as 2022-05-09 IMG_7039.HEIC: no such table: photoinfo
Error exporting photo (8CB60E24-6FB5-4092-9E04-7E7C31A09FAC: IMG_0636.HEIC) as 2022-07-19 IMG_0636.HEIC: no such table: photoinfo
Error exporting photo (EC88162C-7977-4D3F-8F7C-E52B5473E3BB: IMG_7678.HEIC) as 2022-06-14 IMG_7678.HEIC: no such table: photoinfo
Error exporting photo (B3435243-A378-40E4-8741-86F3812423FE: IMG_7220.HEIC) as 2022-05-30 IMG_7220.HEIC: no such table: photoinfo
Error exporting photo (423C32EE-A5B9-41F0-A569-CDE4FB4C1CC5: IMG_7974.HEIC) as 2022-06-19 IMG_7974.HEIC: no such table: photoinfo
Exporting 7121 photos ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━   0% 0:27:32

Aborted!
^C

unm2@xM1 ~ %  /Users/unm2/.local/bin/osxphotos exportdb --migrate /Volumes/media2/Photos/unm2PhonePhotos/Exported/.osxphotos_export.db
Export database /Volumes/media2/Photos/unm2PhonePhotos/Exported/.osxphotos_export.db is already at latest version 7.1
unm2@xM1 ~ % 
```

Probably that missing table error should bail out as well.

**Comment by RhetTbull** [comment]
I've not been able to recreate this but I think I understand the issue. I believe the database migration was interrupted (either an error occurred or osxphotos was interrupted) during the database migration which left the database in  a partially migrated state.  The best way to do this would be to capture the migrations in a transactions and roll back if there was an error but due to the way the sqlite interface in python is implemented this isn't terribly easy and would require a fair amount of rework in the export db code.  For the time being, I think I've fixed the migration issue in [v0.51.8](https://github.com/RhetTbull/osxphotos/releases/tag/v0.51.8) so if you upgrade, you should be able to do `osxphotos exportdb /path/to/export/db --migrate` to force a successful migration which should then allow your export to work.

**Comment by RhetTbull** [comment]
Closing until this comes back up agian. Not been able to replicate and I think it's been fixed.
